### ISSUE TITLE
About VAELossLayer

### ISSUE THREAD

**Issue by Haliko-ai** [tatsy/keras-generative#16]
Hello,

My question may seem a little simple but I couldn't figure out some parts. I'd appreciate if you give me an explanation.

You define vae_loss as below using a class. As far as I understand, the return of this class is a tensor of x_true.
_vae_loss = VAELossLayer()([x_true, x_pred, z_avg, z_log_var])_

Regarding the loss, shouldn't we use lossfun or add_loss instead of zero_loss? I didn't understand how the model will train with a zero_loss tensor?
_self.vae_trainer = Model(inputs=[x_true], outputs=[vae_loss])
self.vae_trainer.compile(loss=[zero_loss]..._

And this is the class structure. It's like we don't use lossfun or add_loss anywhere in the code.

class VAELossLayer(Layer):
    __name__ = 'vae_loss_layer'

    def __init__(self, **kwargs):
        self.is_placeholder = True
        super(VAELossLayer, self).__init__(**kwargs)

    def lossfun(self, x_true, x_pred, z_avg, z_log_var):
        rec_loss = K.mean(K.square(x_true - x_pred))
        kl_loss = K.mean(-0.5 * K.sum(1.0 + z_log_var - K.square(z_avg) - K.exp(z_log_var), axis=-1))
        return rec_loss + kl_loss

    def call(self, inputs):
        x_true = inputs[0]
        x_pred = inputs[1]
        z_avg = inputs[2]
        z_log_var = inputs[3]
        loss = self.lossfun(x_true, x_pred, z_avg, z_log_var)
        self.add_loss(loss, inputs=inputs)

        return x_true

**Comment by tatsy** [comment]
Hi @Haliko-ai,

Since this code is a bit old and is not maintained for a bit long time, I do not remember the reason very well. But if my memory is correct, this is a workaround to use custom loss function with `compile` and `fit`.

Maybe, when you use `compile` and `fit` the arguments for the loss function must be `y_true` and `y_pred`. But in this case, those for the loss function is not in this form. So, I defined the different loss function layer and add the value of the loss with `add_loss` method.

I think you can refer to the following website for more detail.
https://www.tensorflow.org/guide/keras/train_and_evaluate
